### ISSUE TITLE
Marco is unable to colorize metacity themes

### ISSUE THREAD

**Issue by ghost** [mate-desktop/marco#379]
#### Expected behaviour

themes that inherit gtk colors like gtk:bg[SELECTED] should be recolored by marco

#### Actual behaviour

they are not recolored

#### Steps to reproduce the behaviour

install a theme that recolors the metacity theme and compare it to cinnamon/muffin 

#### MATE general version

1.18-2

#### Package version

Marco 1.18.2-2

#### Linux Distribution

Debian testing

a comparison of Marco and Muffin with a windows 10 theme(https://github.com/sixsixfive/DarK/tree/master/Extras/Sub_themes/DarK-Redmond-10/metacity-1)

![untitled](https://user-images.githubusercontent.com/3780774/35483757-4a4433f4-040b-11e8-8762-a2c814949351.png)

**Comment by chuutatsu** [comment]
Can you point which weirdness are you talking about?

Is DarK supposed to have colored title bar in Mate desktop?

**Comment by sixsixfive** [comment]
>Is DarK supposed to have colored title bar in Mate desktop?

yeas this metacity theme should look like on the second screenshot, it worked with MATE@GTK but now everything is grey:

https://github.com/sixsixfive/DarK/blob/master/Extras/Sub_themes/DarK-Redmond-10/metacity-1/metacity-theme-1.xml#L34

**Comment by raveit65** [comment]
Closing this as reporter is gone (ghost).

**Comment by baer63** [comment]
I can confirm this bug. Marco (running MATE 1.20.4 here) seems to pickup some colors from gtk correctly, but "gtk:bg[SELECTED]" is definitely not working. Just try this metacity theme: https://www.gnome-look.org/p/1007357/ - the title bar should be colored according to the bg[SELECTED] color in whatever gtk theme you are running. It is not.
Marco also seems to have a problem with semi transparent bitmaps in metacity themes. Try this theme: https://www.gnome-look.org/p/1014762/ - it contains semi transparent graphics for the inactive frames. These frames should appear as dark gray semi transparent. They show as black.

**Comment by lukefromdc** [comment]
Those theme colors look like GTK2-only stuff, though I'm not sure of the internals the current css files used to set up GTK3 themes are completely different. Is it the same internally by any chance?

**Comment by baer63** [comment]
I'm not sure if I understand your question correctly, however I do have a matching set of GTK2 & GTK3 themes installed. In fact it is the "Mint-X Aqua" theme. The themes I mentioned contain "metacity-1.xml" files and Marco is still compatible to metacity-1 themes, right? So yes, a syntax like "gtk:bg[SELECTED]" within a metacity-1 theme refers to a gtk2 color which in my case is that bright aqua blue in my theme. That bright blue shows correctly in gtk3 as well as gtk2 apps so I know my matching gtk2 & gtk3 themes are working. But Marco shows some light silver instead of the bright acqua blue it is supposed to pickup from the gtk settings. I don't think that has anything to do with css files. 
As far as I understand Marco has been ported to gtk3. So when Marco is parsing metacity-1 xml files and stumbles across something like "gtk:bg[SELECTED]", it would either have to translate that to the according gtk3 color setting or ask gtk2 (but in that case Marco would have to depend on gtk2 and gtk3 libs at the same time).
BTW, back in the days when Marco was gtk2 based (MATE 1.14), these themes worked perfect.

**Comment by lukefromdc** [comment]
In fact, it is not possible for the same process to load both GTK2 and GTK3, mixed symbols will error out and terminate the process. That's why Firefox had to isolate flash (which is GTK2-only in Linux) to a separate process. Otherwise two different libraries would be providing same-named symbols.

Support for recolorizing themes was lost with the GTK3 transition, to bring it back would have been a scratch rewrite.  There ARE a few window docoration themes (Spidey for instance) which do read the background color, presumably from GTK3 but not sure what they are doing, maybe the more recent Metacity theme format?

**Comment by baer63** [comment]
I beg to differ, but support for recolorizing did **not** get lost during the gtk3 transition and "Spidey" is a perfect example. Change your gtk theme from a light to a dark theme and you'll see Spidey go all dark. That happenes because the Spidey theme references "gtk:bg[NORMAL]" in various spots. Spidey is a classic metacity-1 theme, BTW. So ... "gtk:bg[NORMAL]" gets resolved correctly by Marco whereas "gtk:bg[SELECTED]" does **not**. Spidey itself makes many references to "gtk:bg[SELECTED]" but the "background selection" color never shows in a Spidey rendering.
I repeat: Marco resolves "gtk:bg[NORMAL]" correctly but cannot resolve "gtk:bg[SELECTED]". This is clearly a bug.

**Comment by muktupavels** [comment]
@baer63 can you check if theme works in `metacity`? and/or looks good in `metacity-theme-viewer`?

**Comment by lukefromdc** [comment]
Do the newer Metacity theme formats read from the GTK3 theme and thus use GTK3 syntax instead of GTK2 syntax by any chance?

**Comment by muktupavels** [comment]
Read the code?

https://github.com/mate-desktop/marco/blob/master/src/ui/theme.c#L1258
This is place where gtk colors are parsed from metacity theme format.

Metacity theme format syntax is not directly related to GTK 2 or syntax...

**Comment by lukefromdc** [comment]
I was on my way out the door, with no time to get to that. This code shows the internal color specification inside GTK to still use gtk:fg[NORMAL] etc.   It's the GTK theming that was a total do-over between GTK2 and GTK3. 

Anyway, what we have here is that some metacity themes can be recolored and others cannot be. Obviously some are not intended to be recolored and never were. What themes than are supposed to support recoloring but don't work, which ones support it and do work, and what is the common difference between them?

**Comment by baer63** [comment]
OK, so I did some experimenting with one and the same theme (Spidey), replacing "gtk:bg[SELECTED]" with other colors, like "gtk:bg[NORMAL]" or "gtk:text[NORMAL]". All of these options work and the theme is colored accordingly - with the exception of "gtk:bg[SELECTED]" which does not work! This bug seems to be inherited from metacity because the metacity theme viewer shows the same problem.

I am not the only one who has experienced this bug. Just look at this commit for the Mint-Y theme: https://github.com/linuxmint/mint-y-theme/commit/54e67322cdc033fe92ee7e428c77b1c001ae9adb

They started hard-coding color values instead of using "gtk:bg[SELECTED]". So yes, this seems to be a long standing bug which breaks all metacity themes out there which rely on "gtk:bg[SELECTED]" - which is one of the most important colors of a gtk theme since it's the "accent" color (the background for selected or active items).

Regarding the use of semi-transparent bitmaps ... that's another story and I'll open a separate issue for that (that is a problem which occurs only in Marco - metacity does it right).

**Comment by muktupavels** [comment]
```
widget:selected {
  background: red;
}
```

What gtk version you have? Can you try to add above css to `~/.config/gtk-3.0/gtk.css` and then check if `gtk:bg[SELECTED]` results in red color?

**Comment by baer63** [comment]
> 
> 
> ```
> widget:selected {
>   background: red;
> }
> ```
> 
> What gtk version you have? Can you try to add above css to `~/.config/gtk-3.0/gtk.css` and then check if `gtk:bg[SELECTED]` results in red color?

My gtk version is (Debian) 3.22.11 - I tried your suggestion and the result was - well - astonishing.

So let me show you this step by step. My gtk (2&3) theme is Mint-X-Aqua (from Linux Mint). The 1st screenshot shows Caja with my Downloads folder being "selected". The window decoration is Spidey. You can clearly see the bright blue accent color (for selected stuff). Spidey does not show it. The window frame is some silver color.
![Screenshot at 2019-04-18 22-47-19](https://user-images.githubusercontent.com/49685699/56402890-2d071500-622d-11e9-9d1c-519a4ca79179.png)

Now the big surprise - I added your gtk.css mod. And this is the result:
![Screenshot at 2019-04-18 22-48-55](https://user-images.githubusercontent.com/49685699/56402930-5d4eb380-622d-11e9-86e8-930acd4660fd.png)

So what's going on here? The selection color for the Downloads folder is now red and so is Spidey's window frame. I'm not an expert in gtk3 themes. I tried several other gtk3 themes and they all create the same problem - Marco does not pick up the background color.
But hey, looks like you're the man to help solve this puzzle.

**Comment by lukefromdc** [comment]
`gtk:bg[SELECTED]` should work since anything else does
I just went to 
https://developer.gnome.org/gtk3/stable/gtk3-Standard-Enumerations.html#GtkStateFlags
and the currently supported list of state flags still includes
GTK_STATE_FLAG_SELECTED 
and I have it all over my theme without getting deprecation warnings when running things in terminal so it looks like GTK isn't the problem.

**Comment by muktupavels** [comment]
Well I would say that it just confirms that code is working...

`widget:selected` is too generic here and is not real solution. What happens here is that widget / style context that is used to get color is missing `xxx:selected` in theme. What that means is that it is receiving normal background color (gtk:bg[NORMAL] == gtk:bg[SELECTED]). Well of course I could be wrong, it is long time since last time I touched theme related code...

With `metacity` and `metacity-theme-viewer` you should / might be able to get better or even correct result with `window:selected { xxx }` or `window.background:selected { xxx }`. Even if you don't use `metacity`, maybe you can confirm if it works or not?

In metacity `GtkStyleContext` is no longer received from `MetaFrames` widget so things works differently in marco as it still use style from `MetaFrames`. It is subclass from `GtkInvisible` and I think it does not have any special name or anything else that you could use in theme. I see that marco adds background class to style widget. So best you can currently use is `widget.background:selected {}`, I think. Still too generic in my opinion...

**Comment by lukefromdc** [comment]
That suggests maybe adding a widget name to the frame?

**Comment by muktupavels** [comment]
You can try... Of course after that you would need update GTK 3 themes to add needed style.

**Comment by lukefromdc** [comment]
We only control our own themes, meaning this won't fix 3ed party themes unless used inside Marco to read all themes and include the resulting color with that widget name in the style provider.

**Comment by muktupavels** [comment]
https://developer.gnome.org/gtk3/stable/GtkWidget.html#gtk-widget-class-set-css-name
I would use this not widget name... But both should work and both will require updating themes that cares about MATE.

**Comment by baer63** [comment]
@muktupavels I can confirm that with metacity I can add a "window:selected { xxx }" or even a "window.background:selected { xxx }" to my gtk.css and everything works fine. Marco, however, is ignorant of all these hints.

**Comment by baer63** [comment]
@muktupavels For now I am running this code in my ~/.config/gtk-3.0/gtk.css:
```
widget:selected  {
  background: @theme_selected_bg_color;
}
```
It seems to fix the problems at least with the gtk3 themes I am using. I would suggest, however, that Marco should pickup the theme's general selected background color instead of the widget class' selected background color - if that's possible. Seems to make much more sense to me.

**Comment by muktupavels** [comment]
In theme format 3 it is possible to use named colors in following format:
`gtk:custom(color_name,fallback)`

The fallback there is (I think) for one reason - you have zero guarantees that your theme will have defined that color.

**Comment by baer63** [comment]
OK, so I have made my point, I have shown how I can overcome the Marco bug for the themes I am using. Now the question is: will this issue be reopened so that it will be taken care of or did we just have a discussion to waste some time? Most users probably don't even know that many themes don't even look like they are supposed to look. Thanks again @muktupavels for getting to the bottom of this problem.

**Comment by lukefromdc** [comment]
Reopening due to new report and discussion of possibly fixable cause
